### PR TITLE
Consolidate ZSTD utility functions

### PR DESCRIPTION
**Pull request by pjreed**
The zstd_compressor and zstd_decompressor implementations had a
number of duplicated utility functions between them; this consolidates
them into one file.

Distro A, OPSEC #2893

### CODE REVIEW

**Comment by emersonknapp** [comment]
Is this intended to be a followup, to be merged after https://github.com/ros2/rosbag2/pull/418 ?

**Comment by pjreed** [comment]
Yes.

**Comment by piraka9011** [comment]
@pjreed can you please rebase this on master?

**Comment by pjreed** [comment]
Sure, all done.

File: rosbag2_compression/src/rosbag2_compression/compression_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+// Used as a parameter type in a function that accepts the output of ZSTD_getFrameContentSize.
+using ZstdGetFrameContentSizeReturnType = decltype(ZSTD_getFrameContentSize(nullptr, 0));
+
+/**
+ * Open a file using the OS-specific C API.
+ * \param uri is the path to the file.
+ * \param read_mode is the read mode accepted by OS-specific fopen.
+ * \return the FILE pointer or nullptr if the file was not opened.
+ */
+FILE * open_file(const std::string & uri, const std::string & read_mode);
```
Comment (emersonknapp) [inline rosbag2_compression/src/rosbag2_compression/compression_utils.hpp:57]: Just want to double check, could we hide any of these functions private to the compression_utils compilation unit - or are they all being used elsewhere?

Comment (pjreed) [reply rosbag2_compression/src/rosbag2_compression/compression_utils.hpp:57]: I don't believe any of them are being used elsewhere, although I'm not familiar with what you mean by "hide any of these functions private to the compression_utils compilation unit"; is there an example of that somewhere?

Comment (emersonknapp) [reply rosbag2_compression/src/rosbag2_compression/compression_utils.hpp:57]: By that I mean don't declare them in the header, instead put them in an unnamed namespace in the cpp, e.g.

```
// header.hpp

void my_public_function();
```


```
// impl.cpp
#include "header.hpp"

namespace {
void some_helper_fn() {
}
}

void my_public_fn() {
  some_helper_fn();
}

```

Comment (pjreed) [reply rosbag2_compression/src/rosbag2_compression/compression_utils.hpp:57]: Oh, I see what you mean now.  I checked and `open_file` is actually the only one that's only being used in compression_utils, but I'll move it into the cpp file.

**Comment by emersonknapp** [comment]
@emersonknapp - please run this CI job
Gist: https://gist.githubusercontent.com/emersonknapp/0e911aeac5e05944aaf1a9705d652461/raw/99d3e3409ab23ba390808802d886355f5ac19096/ros2.repos
BUILD args: --packages-up-to rosbag2_compression
TEST args: --packages-select rosbag2_compression
Job: ci_launcher

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11564/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6718/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9455/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11531/)

**Comment by emersonknapp** [comment]
Well, there is a single compiler warning on Windows, but it's in a source file you didn't touch, so LGTM

**Comment by emersonknapp** [comment]
It was from an unused variable that was fixed in https://github.com/ros2/rosbag2/pull/460

**Pull request merged**